src/psyclone/tests/psyir/backend/fortran_test.py

```diff
@@ -267,19 +268,30 @@ def test_fw_gen_vardecl(fort_writer):
     assert result == "integer :: dummy1\n"
 
     # Array with intent
-    symbol = DataSymbol("dummy2", "integer", shape=[2, None, 2],
+    symbol = DataSymbol("dummy2", "integer",
+                        shape=[2, DataSymbol.Extent.ATTRIBUTE, 2],
                         interface=ArgumentInterface(
                             ArgumentInterface.Access.READ))
     result = fort_writer.gen_vardecl(symbol)
     assert result == "integer, dimension(2,:,2), intent(in) :: dummy2\n"
 
     # Array with unknown intent
-    symbol = DataSymbol("dummy2", "integer", shape=[2, None, 2],
+    symbol = DataSymbol("dummy2", "integer",
+                        shape=[2, DataSymbol.Extent.ATTRIBUTE, 2],
                         interface=ArgumentInterface(
                             ArgumentInterface.Access.UNKNOWN))
     result = fort_writer.gen_vardecl(symbol)
     assert result == "integer, dimension(2,:,2) :: dummy2\n"
 
+    # Allocatable array
+    symbol = DataSymbol("dummy2", "real", shape=[DataSymbol.Extent.DEFERRED,
+                                                 DataSymbol.Extent.DEFERRED],
+                        interface=ArgumentInterface(
+                            ArgumentInterface.Access.READWRITE))
+    result = fort_writer.gen_vardecl(symbol)
+    assert result == \
+        "real, allocatable, dimension(:,:), intent(inout) :: dummy2\n"
+
```

Comment:
Since it is 'at least 1 DEFERRED' to be an allocatable maybe it will be good to have a second test array with something like "shape=[2, DataSymbol.Extent.DEFERRED, 2]"?

Reply:
The 'at least 1 DEFERRED' was to cover languages other than Fortran (an allocatable array in Fortran must have every dimension specified as ':'). I've added a check for this in the visitor plus associated test.

Reply:
Actually, the situation is more complicated than that - arrays that are dummy arguments may be of 'assumed size' where the extents of one or more of the outer ranks of an array are unspecified, e.g.:

    integer, intent(in) :: my_array(3,:)

Reply:
An array that is a dummy argument may be 'assumed shape' in which case all ranks are specified with ':' and the extents are taken from the actual argument. This actual array *need not* be allocatable.

Reply:
Happily, in both of those cases, the frontend will correctly store the unspecified extents as 'ATTRIBUTE' because there's no 'allocatable' attribute on the declaration.

Reply:
I've added an additional check (in the backend) that we are about to generate a valid assumed-size declaration.